tests/paramgrill.c

```diff
@@ -340,76 +891,264 @@ static int BMK_seed(winnerInfo_t* winners, const ZSTD_compressionParameters para
     return better;
 }
 
+/* bounds check in sanitize too? */
+#define CLAMP(var, lo, hi) {                          \
+    var = MAX(MIN(var, hi), lo);                      \
+}
 
 /* nullified useless params, to ensure count stats */
-static ZSTD_compressionParameters* sanitizeParams(ZSTD_compressionParameters params)
+/* no point in windowLog < chainLog (no point 2x chainLog for bt) */
+/* now with built in bounds-checking */
+/* no longer does anything with sanitizeVarArray + clampcheck */
+static ZSTD_compressionParameters sanitizeParams(ZSTD_compressionParameters params)
 {
-    g_params = params;
     if (params.strategy == ZSTD_fast)
         g_params.chainLog = 0, g_params.searchLog = 0;
     if (params.strategy == ZSTD_dfast)
         g_params.searchLog = 0;
-    if (params.strategy != ZSTD_btopt && params.strategy != ZSTD_btultra)
+    if (params.strategy != ZSTD_btopt && params.strategy != ZSTD_btultra && params.strategy != ZSTD_fast)
         g_params.targetLength = 0;
-    return &g_params;
+
+    return params;
 }
 
+/* new length */
+/* keep old array, will need if iter over strategy. */
+static int sanitizeVarArray(const int varLength, const U32* varArray, U32* varNew, const ZSTD_strategy strat) {
+    int i, j = 0;
+    for(i = 0; i < varLength; i++) {
+        if( !((varArray[i] == CLOG_IND && strat == ZSTD_fast)
+            || (varArray[i] == SLOG_IND && strat == ZSTD_dfast) 
+            || (varArray[i] == TLEN_IND && strat != ZSTD_btopt && strat != ZSTD_btultra && strat != ZSTD_fast))) {
+            varNew[j] = varArray[i];
+            j++;
+        }
+    }
+    return j;
+
+}
 
-static void paramVariation(ZSTD_compressionParameters* ptr)
+/* res should be NUM_PARAMS size */
+/* constructs varArray from ZSTD_compressionParameters style parameter */
+static int variableParams(const ZSTD_compressionParameters paramConstraints, U32* res) {
+    int j = 0;
+    if(!paramConstraints.windowLog) {
+        res[j] = WLOG_IND;
+        j++;
+    }
+    if(!paramConstraints.chainLog) {
+        res[j] = CLOG_IND;
+        j++;
+    }
+    if(!paramConstraints.hashLog) {
+        res[j] = HLOG_IND;
+        j++;
+    }
+    if(!paramConstraints.searchLog) {
+        res[j] = SLOG_IND;
+        j++;
+    }
+    if(!paramConstraints.searchLength) {
+        res[j] = SLEN_IND;
+        j++;
+    }
+    if(!paramConstraints.targetLength) {
+        res[j] = TLEN_IND;
+        j++;
+    }
+    return j;
+}
+
+/* amt will probably always be \pm 1? */
+/* slight change from old paramVariation, targetLength can only take on powers of 2 now (999 ~= 1024?) */
```

Comment:
issue is, experience showed we need intermediate `targetLength` values (non power of 2) to properly scale compression levels.
You might consider a kind of "fixed point" representation, with just 1 or 2 bits representing fractional, so that it becomes possible to test values such as `48` or `24`.